packages/dev/core/src/Maths/math.vector.ts

```diff
@@ -2515,6 +2515,21 @@ export class Vector3 {
         return result;
     }
 
+    /**
+     * Reflects a vector off the plane defined by a normal
+     * @param inDirection defines the vector direction
+     * @param normal defines the normal
+     * @returns the resulting vector
+     */
+    public static Reflect<T extends Vector3>(inDirection: DeepImmutable<Vector3>, normal: DeepImmutable<Vector3>): T {
+        const _inDirection = new (inDirection.constructor as Vector3Constructor<T>)();
+        const _normal = new (normal.constructor as Vector3Constructor<T>)();
+        _inDirection.copyFrom(inDirection);
+        _normal.copyFrom(normal).normalize();
+
+        return _inDirection.subtract(_normal.scale(2 * Vector3.Dot(_inDirection, _normal)));

```

Comment:
You should avoid creating too many temporaries to limit GC. Also, you should add a `toRef` variant, so something like:

```typescript
public static Reflect<T extends Vector3>(inDirection: DeepImmutable<Vector3>, normal: DeepImmutable<Vector3>): Vector3 {
    return this.ReflectToRef(inDirection, normal, new Vector3());
}

public static ReflectToRef<T extends Vector3>(inDirection: DeepImmutable<Vector3>, normal: DeepImmutable<Vector3>, ref: T): T {
    const tmp = TmpVectors.Vector3[0];
    tmp.copyFrom(normal).scaleInPlace(2 * Vector3.Dot(inDirection, normal));

    return ref.copyFrom(inDirection).subtractInPlace(tmp);
}
```